### PR TITLE
Generic Entities Interface

### PR DESCRIPTION
**Pull request by j-forster**
Implemented the generic entities interface.

### CODE REVIEW

**Comment by j-forster** [comment]
typo in https://github.com/Waziup/api-server/compare/master...j-forster:master#diff-e51481b1460633a96956b4acb868ba4bR79
should be 'deleteEntityMeasMongo'

**Comment by j-forster** [comment]
Please see the changes at https://github.com/j-forster/api-server/commit/a03d8b555879155121475ba39752de4599e0b42e#diff-9d9a6cd82f41984872a66a3ab0d440c4
Environment variable changed from 'SERVER_URL' to 'HTTP_URL' and 'SERVER_PORT' to 'HTTP_PORT'.
Also the default ports are different from the well known ports.
'HTTPS_ENABLED' is false by default, 'HTTP_ENABLED' is true.

File: src/routes/entities/entities.route.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+const orionProxy = require('../../lib/orion-proxy');
+
+const orionRequest = orionProxy.orionRequest;
+
+
+async function createEntity(domain, body) {
+  
+  var entity = {
+    id: body.id,
+    type: body.type
```
Comment (cdupont) [inline src/routes/entities/entities.route.js:13]: Why the rest of the entity is not saved? for instance the attributes

File: src/routes/entities/entities.route.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+
+async function getEntityTypes(domain) {
+  
+  var resp = orionRequest('/v2/types?options=values', 'GET', domain);
+  return resp;
+}
+
+async function getEntities(domain, type, body) {
+  
+  var entities  = await orionRequest('/v2/entities?type='+type+'&attrs=id', 'GET', domain);
```
Comment (cdupont) [inline src/routes/entities/entities.route.js:28]: Should you call mongo as well? To complete the entity with historical values...

File: src/routes/entities/entities.route.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+      entity[key] = orionEntity[key].value;
+    }
+    
+    return entity
+  }
+}
+
+async function deleteEntity(domain, type, id) {
+  
+  var resp = orionRequest('/v2/entities/'+id+'?type='+type, 'DELETE', domain);
```
Comment (cdupont) [inline src/routes/entities/entities.route.js:56]: Similarly, should you delete the data from Mongo?

File: src/routes/entities/entities.route.js [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+async function getEntityAttribute(domain, type, id, attr) {
+  
+  await orionRequest('/v2/entities/'+id+'/attrs/'+attr+'?type='+type, 'GET', domain);
+  var measurements = await mongoProxy.getEntityMeasurementValues(domain, id, type, attr);
+  return measurements.map(meas => meas.value);
+}
+
+async function deleteEntityAttribute(domain, type, id, attr) {
+  
+  var resp = orionRequest('/v2/entities/'+id+'/attrs/'+attr+'?type='+type, 'DELETE', domain);
```
Comment (cdupont) [inline src/routes/entities/entities.route.js:69]: Should you delete from mongo?

File: src/routes/entities/entities.route.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+  return resp;
+}
+
+async function putEntityAttribute(domain, type, id, attr, body) {
+  
+  await mongoProxy.deleteEntityMeasMongo(domain, id, type, attr);
```
Comment (cdupont) [inline src/routes/entities/entities.route.js:75]: This line might be the typo

**Review by cdupont — COMMENTED** [review]
It's not clear for me why the entities endpoint needs Mongo? Mongo was used in the Sensors case because it allows to store the history of measurement values. In the case of generic entities, we cannot know if the user wants to store historical data and which one...
All in all, I wouldn't use Mongo in the generic entities.

**Comment by j-forster** [comment]
> It's not clear for me why the entities endpoint needs Mongo? Mongo was used in the Sensors case because it allows to store the history of measurement values. In the case of generic entities, we cannot know if the user wants to store historical data and which one...
All in all, I wouldn't use Mongo in the generic entities.

With `PUT ../entities/Room/myRoom/temperature` and `GET ../entities/Room/myRoom/temperature`
the user never needs to care about historical values.

If the user wants to have historical data, he doesnt need a special endpoint, he can just use:
`POST ../entities/Room/myRoom/temperature` and `GET ../entities/Room/myRoom/temperature`

See the swagger specification:
Use POST to insert values to the historical values.
Use PUT to set the value like a single value without historical values.
Use GET to get the value or values.

I am using MongoDB because you are using it. You can get all sensor data (from the /sensors enpoints) with the /entities enpoints.
Just call `GET ../entities/SensingDevice/Sensor2-../temperature`

File: src/routes/entities/entities.route.js (hunk 1 above)
Comment (j-forster) [reply src/routes/entities/entities.route.js:13]: Creating entities and managing attributes is separated!
Use `POST ../entities` to create a new (blank) entity and
use `POST PUT GET ../entities/<Type>/<ID>/<Attr>` to create or modify the attributes.

File: src/routes/entities/entities.route.js (hunk 2 above)
Comment (j-forster) [reply src/routes/entities/entities.route.js:28]: The `GET ../entities/Room/myRoom` returns just the current value.
If you are interested in historical values,
use `GET ../entities/Room/myRoom/temperature` with the specific attribute.
This should resolve the problem, that a entity with a lot of attributes and 1000 values for each attribute will generate a huge `GET` request, even if just just want to get the basic entity structure.

File: src/routes/entities/entities.route.js (hunk 3 above)
Comment (j-forster) [reply src/routes/entities/entities.route.js:56]: This is done by the proxy: https://github.com/j-forster/api-server/blob/master/src/proxy.js#L99
At the /sensors enpoints, you did exactly the same: https://github.com/j-forster/api-server/blob/master/src/proxy.js#L66

File: src/routes/entities/entities.route.js (hunk 4 above)
Comment (j-forster) [reply src/routes/entities/entities.route.js:69]: Here: https://github.com/j-forster/api-server/blob/master/src/proxy.js#L108

File: src/routes/entities/entities.route.js (hunk 5 above)
Comment (j-forster) [reply src/routes/entities/entities.route.js:75]: No 😆 
The POST request *adds* values to the historical values, but 
the PUT requests *sets* the value and *deletes all historical values*.

Call PUT and GET to get and set single values.
Call POST and GET to set values and have them stored persistent and query them all later.

File: src/proxy.js [hunk 6]
```diff
@@ -75,14 +76,38 @@ function installSensors(router, keycloak) {
                                                                                        proxy(req => mongoProxy.postSensorMeasMongo(       req.params.domain, req.params.sensorID, req.body), true));
   router.get(    '/domains/:domain/sensors/:sensorID/measurements/:measID',            proxy(req => orionProxy.getSensorMeasurement(      req.params.domain, req.params.sensorID, req.params.measID), true));
   router.delete( '/domains/:domain/sensors/:sensorID/measurements/:measID',            proxy(req => orionProxy.deleteSensorMeasurement(   req.params.domain, req.params.sensorID, req.params.measID)),
-                                                                                       proxy(req => mongoProxy.deleteMeasMongo(           req.params.domain, req.params.sensorID, req.params.measID), true));
+                                                                                       proxy(req => mongoProxy.deleteMeasMongo(           req.params.domain, req.params.sensorID, "SensingDevice", req.params.measID), true));
```
Comment (j-forster) [inline src/proxy.js:79]: @cdupont Thats the typo: 'deleteEntityMeasMongo' not 'deleteMeasMongo'. I will fix it soon.

**Pull request merged**